src/bootstrap/compile.rs

```diff
@@ -517,6 +517,95 @@ impl Step for StartupObjects {
     }
 }
 
+#[derive(Debug, PartialOrd, Ord, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct CrateLibrustc {}
+impl Step for CrateLibrustc {
+    type Output = ();
+
+    fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
+        run.krate("rustc-main")
+    }
+
+    fn make_run(run: RunConfig<'_>) {
+        let krate = run.builder.crate_paths[&run.path];
+        run.builder.ensure(Crate {
+            target: run.target,
+            mode: Mode::Rustc,
+            krate,
+        });
+    }
+
+    fn run(self, _: &Builder<'_>) {
+        unreachable!()
+    }
+}
+
+/// Build a single crate using cargo.
+///
+/// FIXME: once test flags are unified under `builder.cargo`,
+/// unify this with `test::Crate`.
+#[derive(Debug, PartialOrd, Ord, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct Crate {
+    pub target: TargetSelection,
+    pub mode: Mode,
+    pub krate: Interned<String>,
+}
+
+impl Step for Crate {
+    type Output = ();
+
+    fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
+        run.krate("test")
+    }
+
+    fn make_run(run: RunConfig<'_>) {
+        let krate = run.builder.crate_paths[&run.path];
+        run.builder.ensure(Crate {
+            target: run.target,
+            // NOTE: ok to hard-code `Std`, even though `CrateRustc` uses `ensure(Crate)`,
+            // because `make_run` is only called for dependencies of `test`.
+            mode: Mode::Std,
+            krate,
+        });
+    }
+
+    fn run(self, builder: &Builder<'_>) {
+        let Crate { mode, target, krate } = self;
+
+        let compiler = builder.compiler_for(builder.top_stage, builder.config.build, target);
+        let mut cargo =
+            builder.cargo(compiler, mode, SourceType::InTree, target, builder.kind.as_str());
+        let stamp = match mode {
+            Mode::Std => {
+                std_cargo(builder, target, compiler.stage, &mut cargo);
+                libstd_stamp(builder, compiler, target)
+            }
+            Mode::Rustc => {
+                builder.ensure(Std { compiler, target });
+                rustc_cargo(builder, &mut cargo, target);
+                librustc_stamp(builder, compiler, target)
+            }
+            _ => panic!("crate-specific builds only supported for library and compiler"),
+        };
+
+        cargo.arg("-p").arg(krate);
+
+        builder.info(&format!(
+            "{} {} stage{} ({} -> {})",
+            builder.kind.as_str(), krate, compiler.stage, &compiler.host, target
+        ));
+
+        run_cargo(
+            builder,
+            cargo,
+            vec![],
+            &stamp,
+            vec![],
+            false,
+        );
+    }
```

Comment:
hmm, in retrospect I think all this is wrong, because it doesn't include the logic from `Std::run` and `Rustc::run` :/ I can move those out into functions maybe? have to think about this more.

Reply:
Ok, how about this - I can change `Std` to use `krates()` instead of `all_krates()` and add a bunch of `-p core` arguments to the `tail_args` that get passed to `run_cargo` here. That should have exactly the same effect without having to introduce much code at all. The only downside is it requires adding a bunch of `tail_args: vec![]` calls all over the place where `Std` is constructed.